ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,490 @@
+        epoll_fd = epoll_create(1)
+        assert(epoll_fd != -1)
+        ev.data.fd = ucp_epoll_fd
+        ev.data.ptr = NULL
+        ev.data.u32 = 0
+        ev.data.u64 = 0
+        ev.events = EPOLLIN
+        err = epoll_ctl(epoll_fd, EPOLL_CTL_ADD, ucp_epoll_fd, &ev)
+        assert(err == 0)
+        return epoll_fd
+
+    def arm(self):
+        cdef ucs_status_t status
+        status = ucp_worker_arm(self._handle)
+        if status == UCS_ERR_BUSY:
+            return False
+        assert_ucs_status(status)
+        return True
+
+    def ep_create(self, str ip_address, port):
+        cdef ucp_ep_params_t params
+        ip_address = socket.gethostbyname(ip_address)
+        if c_util_get_ucp_ep_params(&params, ip_address.encode(), port):
+            raise MemoryError("Failed allocation of ucp_ep_params_t")
+
+        cdef ucp_ep_h ucp_ep
+        cdef ucs_status_t status = ucp_ep_create(self._handle, &params, &ucp_ep)
+        c_util_get_ucp_ep_params_free(&params)
+        assert_ucs_status(status)
+        return ucx_ep_create(ucp_ep)
+
+    def progress(self):
+        while ucp_worker_progress(self._handle) != 0:
+            pass
+
+    @property
+    def handle(self):
+        return int(<uintptr_t><void*>self._handle)
+
+    def request_cancel(self, ucp_request_as_int):
+        cdef ucp_request *req = <ucp_request*><uintptr_t>ucp_request_as_int
+
+        # Notice, `ucp_request_cancel()` calls the send/recv callback function,
+        # which will handle the request cleanup.
+        ucp_request_cancel(self._handle, req)
+
+
+cdef class UCXEndpoint:
+    """Python representation of `ucp_ep_h`
+    Please use `ucx_ep_create()` to contruct an instance of this class
+    """
+    cdef:
+        ucp_ep_h _handle
+        bint initialized
+
+    def __cinit__(self):
+        self._handle = NULL
+        self.initialized = False
+
+    def close(self, UCXWorker worker):
+        cdef ucs_status_ptr_t status
+        if self.initialized:
+            status = ucp_ep_close_nb(self._handle, UCP_EP_CLOSE_MODE_FLUSH)
+            self.initialized = False
+            if UCS_PTR_STATUS(status) != UCS_OK:
+                assert not UCS_PTR_IS_ERR(status)
+                # We spinlock here until `status` has finished
+                while ucp_request_check_status(status) != UCS_INPROGRESS:
+                    worker._worker.progress()
+                assert not UCS_PTR_IS_ERR(status)
+                ucp_request_free(status)
+
+    def info(self):
+        assert self.initialized
+        # Making `ucp_ep_print_info()` write into a memstream,
+        # convert it to a Python string, clean up, and return string.
+        cdef char *text
+        cdef size_t text_len
+        cdef FILE *text_fd = open_memstream(&text, &text_len)
+        assert(text_fd != NULL)
+        ucp_ep_print_info(self._handle, text_fd)
+        fflush(text_fd)
+        cdef unicode py_text = text.decode()
+        fclose(text_fd)
+        free(text)
```

Comment:
Should we have some `try...finally...` block above to make sure we cleanup if an error comes up in the intervening code?

Reply:
Fixed